docs/node-mixin/rules/rules.libsonnet

```diff
@@ -0,0 +1,116 @@
+{
+  prometheusRules+:: {
+    groups+: [
+      {
+        name: 'node-exporter.rules',
+        rules: [
+          {
+            // This rule gives the number of CPUs per node.
+            record: 'instance:node_num_cpu:sum',
+            expr: |||
+              count without (cpu) (
+                sum without (mode) (
+                  node_cpu_seconds_total{%(nodeExporterSelector)s}
+                )
+              )
+            ||| % $._config,
+          },
+          {
+            // CPU utilisation is % CPU is not idle.
+            record: 'instance:node_cpu_utilisation:avg_rate1m',
+            expr: |||
+              1 - avg without (cpu, mode) (
+                rate(node_cpu_seconds_total{%(nodeExporterSelector)s, mode="idle"}[1m])
+              )
+            ||| % $._config,
+          },
+          {
+            // This is CPU saturation: 1min avg run queue length / number of CPUs.
+            // Can go over 1. >1 is bad.
+            record: 'instance:node_load1_per_cpu:ratio',
+            expr: |||
+              (
+                node_load1{%(nodeExporterSelector)s}
+              /
+                instance:node_num_cpu:sum{%(nodeExporterSelector)s}
+              )
+            ||| % $._config,
+          },
+          {
+            // Memory utilisation per node, normalized by per-node memory
+            record: 'instance:node_memory_utilisation:ratio',
+            expr: |||
+              1 - (
+                node_memory_MemAvailable_bytes{%(nodeExporterSelector)s}
+              /
+                node_memory_MemTotal_bytes{%(nodeExporterSelector)s}
+              )
+            ||| % $._config,
+          },
+          {
+            record: 'instance:node_memory_swap_io_pages:rate1m',
+            expr: |||
+              (
+                rate(node_vmstat_pgpgin{%(nodeExporterSelector)s}[1m])
+              +
+                rate(node_vmstat_pgpgout{%(nodeExporterSelector)s}[1m])
+              )
+            ||| % $._config,
+          },
+          {
+            // Disk utilisation (seconds spent, 1 second rate)
+            record: 'instance:node_disk_io_time:sum_rate1m',
```

Comment:
the sum of a rate is just :rate, and you've lost some of the metric name.

This should be instance:node_disk_io_time_seconds:rate1m

There's also no reason to restrict this to only some devices, if you want to limit alerts/graphs do so there rather than having a recording rule produce partial data.


Similar comments below.

Reply:
I'll leave the `diskDeviceSelector` in, but it will be empty as the default value (no device selection be default).
I understand the `sum` here as description that the rate of each device is added up rather than averaged or maximum. 
(Which makes me think if `sum` is actually the best aggregation here. For CPU utilization, we took the average and for the stacked utilization graph even took into account how many cores each instance has. Instances with multiple disks fully utilized will get values >1 in this sum, and that's not accounted for. `avg` would help with that, but it would let the utilization appear artificially low if an instance has an idle secondary disk sitting around. Perhaps treat all devices separate and don't even attempt to aggregate per instance? I think I have to play with this a bit to come to a good conclusion. Once more, I'll add a TODO here.)

Reply:
I changed the name from `instance:node_disk_io_time:sum_rate1m` to `instance:node_disk_io_time_seconds:sum_rate1m` and similar for `instance:node_disk_io_time_weighted:sum_rate1m`. I'll leave the `sum` in for now as I think it carries meaning. (But obviously, should we go for not aggregating together devices, there will be no `sum` in the aggregation and no `sum` in the rule name.)

Reply:
I'll work on per-device rules tomorrow. I think that's important to get right in this PR.

Reply:
Sum is the default aggregation, so it adds little value. Depending on how you go about an aggregation you could have sum or sum_sum_sum as additions are associative. I've come across a case before where one group decided to ignore best practice and sometimes include sum, it was messy, so I'd like to avoid that.

Reply:
OK, `sum` as default (like `job` for labels), only need it if it is the only operation. I think I understand. Will rework with this in mind.